src/infra/aws/client/kms.go

```diff
@@ -116,25 +140,63 @@ func (c *AWSClient) Sign(_ context.Context, keyID string, msg []byte, signingAlg
 	return out, nil
 }
 
-func (c *AWSClient) DeleteKey(_ context.Context, keyID string) (*kms.ScheduleKeyDeletionOutput, error) {
+func (c *AWSClient) DeleteKey(ctx context.Context, keyID string) (*kms.ScheduleKeyDeletionOutput, error) {
 	out, err := c.kmsClient.ScheduleKeyDeletion(&kms.ScheduleKeyDeletionInput{
 		KeyId: &keyID,
 	})
 	if err != nil {
 		return nil, parseKmsErrorResponse(err)
 	}
 
+	err = c.waitKeyState(ctx, keyID, func(metadata *kms.KeyMetadata) error {
+		if !*metadata.Enabled && *metadata.KeyState == KeyStatePendingDeletion {
+			return nil
+		}
+
+		return fmt.Errorf("key %s was not scheduled for deletion yet", keyID)
+	})
+	if err != nil {
+		return nil, err
+	}
+
 	return out, nil
 }
 
-func (c *AWSClient) RestoreKey(_ context.Context, keyID string) (*kms.CancelKeyDeletionOutput, error) {
+func (c *AWSClient) RestoreKey(ctx context.Context, keyID string) (*kms.CancelKeyDeletionOutput, error) {
 	out, err := c.kmsClient.CancelKeyDeletion(&kms.CancelKeyDeletionInput{
 		KeyId: &keyID,
 	})
 	if err != nil {
 		return nil, parseKmsErrorResponse(err)
 	}
 
+	err = c.waitKeyState(ctx, keyID, func(metadata *kms.KeyMetadata) error {
+		if *metadata.KeyState == KeyStatePendingDeletion {
+			return fmt.Errorf("key %s is still pending for deletion deletion", keyID)
+		}
```

Comment:
Here it will only be accepted in that case, so we should only accept that state

Reply:
yes, what else should I do instead?

Reply:
This comment was supposed to highlight the misunderstanding above, my mistake. Thats ok